### PR TITLE
feat: support inverted index for array

### PR DESCRIPTION
**Pull request by longjiquan**
issue: https://github.com/milvus-io/milvus/issues/27704

### CODE REVIEW

File: internal/core/src/index/IndexFactory.cpp [hunk 1]
```diff
@@ -255,32 +255,25 @@ IndexFactory::CreateScalarIndex(const CreateIndexInfo& create_index_info,
...
-            return CreateScalarIndex<double>(
-                index_type, file_manager, space, data_type);
+            return CreateScalarIndex<double>(index_type, file_manager, space);
 
             // create string index
         case DataType::STRING:
         case DataType::VARCHAR:
             return CreateScalarIndex<std::string>(
-                index_type, file_manager, space, data_type);
+                index_type, file_manager, space);
```
Comment (zhagnlu) [inline internal/core/src/index/IndexFactory.cpp:276]: V2 not implement array

File: internal/core/src/index/InvertedIndexTantivy.cpp [hunk 2]
```diff
@@ -447,6 +331,176 @@ InvertedIndexTantivy<T>::BuildWithRawData(size_t n,
...
+        }
+
+        case proto::schema::DataType::VarChar: {
+            for (const auto& data : field_datas) {
+                auto n = data->get_num_rows();
+                wrapper_->add_data<std::string>(
+                    static_cast<const std::string*>(data->Data()), n);
+            }
+            break;
+        }
```
Comment (zhagnlu) [inline internal/core/src/index/InvertedIndexTantivy.cpp:409]: why not directly using T replace type
 for (const auto& data : field_datas) {
                auto n = data->get_num_rows();
                wrapper_->add_data<T>(
                    static_cast<const T*>(data->Data()), n);
            }
            break;

File: internal/core/src/index/InvertedIndexTantivy.cpp [hunk 3]
```diff
@@ -447,6 +331,176 @@ InvertedIndexTantivy<T>::BuildWithRawData(size_t n,
...
+
+        default:
+            PanicInfo(ErrorCode::NotImplemented,
+                      fmt::format("Inverted index not supported on {}",
+                                  schema_.data_type()));
+    }
+}
+
+template <typename T>
+template <typename ElementT>
```
Comment (zhagnlu) [inline internal/core/src/index/InvertedIndexTantivy.cpp:487]: for your framework, T and ElementT is same, is that right?  if same, not need ElementT

File: internal/core/src/index/InvertedIndexTantivy.cpp [hunk 4]
```diff
@@ -447,6 +331,176 @@ InvertedIndexTantivy<T>::BuildWithRawData(size_t n,
...
+                                static_cast<DataType>(schema_.element_type()));
+                            std::vector<std::string> output;
+                            for (int64_t j = 0; j < array_column[i].length();
+                                 j++) {
+                                output.push_back(
+                                    array_column[i]
+                                        .template get_data<std::string>(j));
+                            }
+                            wrapper_->template add_multi_data(output.data(),
+                                                              output.size());
```
Comment (zhagnlu) [inline internal/core/src/index/InvertedIndexTantivy.cpp:465]: this code block can wrapper as specify template function for string

Comment (longjiquan) [reply internal/core/src/index/InvertedIndexTantivy.cpp:465]: OK.

File: internal/core/src/index/InvertedIndexTantivy.cpp (hunk 2 above)
Comment (longjiquan) [reply internal/core/src/index/InvertedIndexTantivy.cpp:409]: OK.

File: internal/core/src/index/IndexFactory.cpp (hunk 1 above)
Comment (longjiquan) [reply internal/core/src/index/IndexFactory.cpp:276]: We already depreacted the storage-v2-related api.

File: internal/core/src/index/InvertedIndexTantivy.cpp (hunk 3 above)
Comment (longjiquan) [reply internal/core/src/index/InvertedIndexTantivy.cpp:487]: T and ElementT is not same for the array data type in the first design.

Comment (longjiquan) [reply internal/core/src/index/InvertedIndexTantivy.cpp:487]: And yes they are same now.

File: internal/core/src/index/IndexFactory.cpp [hunk 5]
```diff
@@ -152,39 +139,52 @@ IndexFactory::CreateScalarIndex(
...
+                throw SegcoreError(
+                    DataTypeInvalid,
+                    fmt::format("invalid data type to build index: {}",
+                                data_type));
+        }
+    };
+
+    switch (data_type) {
+        case DataType::ARRAY:
+            // or use Array?
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/index/IndexFactory.cpp:183]: Very complicated to use array as template.

**Comment by longjiquan** [comment]
comment fixed at https://github.com/milvus-io/milvus/pull/33452/commits/073e5aa6c417df900a933837d241b6cce058ffe5

**Comment by zhagnlu** [comment]
> comment fixed at [073e5aa](https://github.com/milvus-io/milvus/commit/073e5aa6c417df900a933837d241b6cce058ffe5)

/lgtm

**Comment by longjiquan** [comment]
rerun ut

File: internal/core/thirdparty/tantivy/tantivy-wrapper.h [hunk 6]
```diff
@@ -49,6 +51,15 @@ struct RustArrayWrapper {
         std::cout << ss.str() << std::endl;
     }
 
+    std::set<uint32_t>
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/thirdparty/tantivy/tantivy-wrapper.h:54]: This api is just used in test. `to_set` should never be used in real filtering sceniar.

[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/thirdparty/tantivy/tantivy-wrapper.h:54]: I'll delete this api in the next pr.

**Pull request merged**